File: core/store.go
package core

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// A base is one directory: a git working tree holding five typed layers. Every path an
// agent, a command, or a URI names resolves through a Store, so no caller ever joins a
// layer name onto a root of its own — that is what keeps a base relocatable as one unit
// and what makes confinement a single enforced rule rather than a convention.

// Layer is one typed storage layer of a base.
type Layer string

const (
	// LayerEvents holds one collected document per source per completed local day.
	LayerEvents Layer = "events"
	// LayerIndex holds one point-in-time document per source: the things you have, as
	// opposed to the things that happened. It was called `index` while it also held the
	// derived caches, which made one word mean a layer, a source kind, two rebuildable files,
	// a wiki command, and a wiki page at once.
	LayerIndex Layer = "index"
	// LayerTasks holds task execution traces.
	LayerTasks Layer = "tasks"
	// LayerProjects holds status-bearing intent pages.
	LayerProjects Layer = "projects"
	// LayerWiki holds the flat OKF v0.2 knowledge bundle.
	LayerWiki Layer = "wiki"
)

// Layers is the canonical ordered layer list, consumed by path resolution, scaffolding,
// configuration validation, and every `list` command.
var Layers = []Layer{LayerEvents, LayerIndex, LayerTasks, LayerProjects, LayerWiki}

// ParseLayer converts a user-supplied name into a known layer.
func ParseLayer(value string) (Layer, error) {
	candidate := Layer(strings.ToLower(strings.TrimSpace(value)))
	for _, layer := range Layers {
		if candidate == layer {
			return layer, nil
		}
	}
	return "", fmt.Errorf("unknown layer %q; valid layers: %s", value, LayerNames())
}

// LayerNames renders the canonical layer list for a diagnostic.
func LayerNames() string {
	names := make([]string, 0, len(Layers))
	for _, layer := range Layers {
		names = append(names, string(layer))
	}
	return strings.Join(names, ", ")
}

// Base permissions. A base is created owner-only whether or not it is versioned: git stores
// no directory mode and tracks only the executable bit, so versioning is never a reason to
// relax them. What versioning changes is that fkf stops repairing modes inside a tree that
// git, an editor, and a teammate's clone also own.
const (
	BaseDirMode  os.FileMode = 0o700
	BaseFileMode os.FileMode = 0o600
)

const (
	baseDirMode  = BaseDirMode
	baseFileMode = BaseFileMode
)

// Generated graph filenames live at the base root. They remain outside every typed layer so
// listings cannot confuse rebuildable cache data with collected or authored content.
const (
	GraphFile         = "graph.tsv"
	GraphMetaFile     = "graph.meta.json"
	TaskTraceFile     = "TASKS.md"
	BaseAgentsFile    = "AGENTS.md"
	BaseSkillsDir     = ".agents/skills"
	BaseBinDir        = "bin"
	ConfigFileName    = "fkf.yaml"
	LocalConfigName   = "fkf.local.yaml"
	MarkdownExtension = ".md"
)

// ErrNotAddressable reports a base-relative path outside the published URI grammar. The CLI
// maps it to exit code 2: naming a file a base does not address is a usage error, not a
// failure.
var ErrNotAddressable = errors.New("path is not addressable in a base")

// pageSlugPattern is the path-level half of the Markdown contract. Content validation owns
// frontmatter and links, but Store.Resolve must reject hidden files, backups, and nested pages
// before a generic read (including MCP) can open them.
var pageSlugPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9._-]*$`)

// ErrLayerDisabled reports a request for a layer this base does not enable. It is a distinct
// error because "you turned it off" and "it is empty" are different answers, and a command
// that conflates them teaches the user to ignore both.
type ErrLayerDisabled struct{ Layer Layer }

func (e ErrLayerDisabled) Error() string {
	return fmt.Sprintf("layer %s is disabled in %s; set layers.%s: true to enable it",
		e.Layer, ConfigFileName, e.Layer)
}

// Store is the resolved, immutable layout of one base.
type Store struct {
	root    string
	enabled map[Layer]bool
}

// NewStore resolves a store from a root and the layer activation the base declares. An
// absent entry is disabled, so a hand-written configuration cannot silently enable a layer.
func NewStore(root string, enabled map[Layer]bool) Store {
	store := Store{
		root:    filepath.Clean(ExpandHome(root)),
		enabled: make(map[Layer]bool, len(Layers)),
	}
	for _, layer := range Layers {
		store.enabled[layer] = enabled[layer]
	}
	return store
}

// Root returns the base directory.
func (s Store) Root() string { return s.root }

// Enabled reports whether a layer is activated.
func (s Store) Enabled(layer Layer) bool { return s.enabled[layer] }

// EnabledLayers returns the activated layers in canonical order.
func (s Store) EnabledLayers() []Layer {
	active := make([]Layer, 0, len(Layers))
	for _, layer := range Layers {
		if s.enabled[layer] {
			active = append(active, layer)
		}
	}
	return active
}

// Dir returns the resolved directory of an enabled layer.
func (s Store) Dir(layer Layer) (string, error) {
	// Resolve applies both activation and the same below-root symlink confinement as a file
	// read. Directory enumeration is still a read: following a committed layer-root symlink
	// would otherwise expose filenames and shape outside the base before a later file open failed.
	return s.Resolve(string(layer))
}

// LayerOf returns the layer a base-relative path belongs to, if any. It is how a URI is
// checked against layer activation without every caller re-deriving the first segment.
func (s Store) LayerOf(relative string) (Layer, bool) {
	first, _, _ := strings.Cut(strings.TrimPrefix(relative, "/"), "/")
	for _, layer := range Layers {
		if first == string(layer) {
			return layer, true
		}
	}
	return "", false
}

// Resolve maps a base-relative slash path to an absolute path inside the base. Every read and
// write in fkf goes through it, so it is where the three ways out of a base are refused: a
// path that escapes lexically, a path addressing a disabled layer, and a path outside the
// addressable set.
//
// The addressable set is the published URI grammar and nothing else. Confining to the root
// was not enough on its own: a base IS a git repository, so `.git/config` always exists beside
// the layers, and a user wiring up a source drops a `.env` there — both were readable through
// `fkf read`, and therefore through the ungated MCP `read` tool. `fkf.local.yaml` stays out
// deliberately; it is the machine-local overlay and no URI names it.
func (s Store) Resolve(relative string) (string, error) {
	cleaned, err := CleanRelative(relative)
	if err != nil {
		return "", err
	}
	cleaned = strings.TrimSuffix(cleaned, "/")
	if cleaned == "." {
		return s.root, nil
	}
	switch layer, isLayer := s.LayerOf(cleaned); {
	case isLayer && !s.enabled[layer]:
		return "", ErrLayerDisabled{Layer: layer}
	case !addressableBasePath(cleaned):
		return "", notAddressable(cleaned)
	}
	absolute := filepath.Join(s.root, filepath.FromSlash(cleaned))
	if err := ValidateWithinRoot(s.root, absolute); err != nil {
		return "", err
	}
	return absolute, nil
}

// addressableBasePath is the shared lexical half of Store.Resolve. Collection validates
// relation paths before it has a Base, so keeping the closed grammar here prevents evidence
// that writes successfully but makes the graph unreadable on its next rebuild.
func addressableBasePath(relative string) bool {
	cleaned := strings.TrimSuffix(relative, "/")
	first, _, _ := strings.Cut(cleaned, "/")
	for _, layer := range Layers {
		if first == string(layer) {
			return addressableLayerPath(layer, cleaned)
		}
	}
	// The root graph files are addressable but are not a layer: there is nothing to enable or
	// browse. Both the rows and their integrity metadata resolve through the Store internally.
	return cleaned == GraphFile ||
		cleaned == GraphMetaFile ||
		cleaned == ConfigFileName ||
		cleaned == BaseAgentsFile
}

type relationFileCapabilities struct {
	Fragment bool
	JQ       bool
}

// relationFilePath is the file-node half of the public relation grammar. Store.Resolve also
// admits directories because listings need them; an edge cannot point at a listing. The two
// optional URI suffixes are restricted to the file kinds whose read path implements them.
func relationFilePath(relative string) (relationFileCapabilities, bool) {
	parts := strings.Split(relative, "/")
	switch {
	case relative == ConfigFileName, relative == GraphFile:
		return relationFileCapabilities{}, true
	case relative == BaseAgentsFile:
		return relationFileCapabilities{Fragment: true}, true
	case relative == GraphMetaFile:
		return relationFileCapabilities{JQ: true}, true
	case len(parts) == 3 && parts[0] == string(LayerEvents) && validAddressDate(parts[1]):
		if addressableSourceDocument(parts[2]) {
			return relationFileCapabilities{Fragment: true, JQ: true}, true
		}
	case len(parts) == 2 && parts[0] == string(LayerIndex) && addressableSourceDocument(parts[1]):
		return relationFileCapabilities{Fragment: true, JQ: true}, true
	case len(parts) == 4 && parts[0] == string(LayerTasks) && validAddressDate(parts[1]) &&
		pageSlugPattern.MatchString(parts[2]) && parts[3] == TaskTraceFile:
		return relationFileCapabilities{Fragment: true}, true
	case len(parts) == 2 && (parts[0] == string(LayerProjects) || parts[0] == string(LayerWiki)):
		slug, found := strings.CutSuffix(parts[1], MarkdownExtension)
		if found && pageSlugPattern.MatchString(slug) {
			return relationFileCapabilities{Fragment: true}, true
		}
	}
	return relationFileCapabilities{}, false
}

func notAddressable(relative string) error {
	return fmt.Errorf("%w: %s (a base addresses the published shapes under the %s layers, plus %s, %s, %s and %s)",
		ErrNotAddressable, relative, LayerNames(), GraphFile, GraphMetaFile, ConfigFileName, BaseAgentsFile)
}

// addressableLayerPath is deliberately a closed grammar. A layer directory may contain
// malformed or private neighbour files because a base is an ordinary git repository; being
// below an enabled layer never grants those files a URI.
func addressableLayerPath(layer Layer, relative string) bool {
	parts := strings.Split(relative, "/")
	if len(parts) == 1 {
		return true
	}
	switch layer {
	case LayerEvents:
		if len(parts) == 2 {
			return validAddressDate(parts[1])
		}
		if len(parts) != 3 || !validAddressDate(parts[1]) {
			return false
		}
		return addressableSourceDocument(parts[2])
	case LayerIndex:
		return len(parts) == 2 && addressableSourceDocument(parts[1])
	case LayerTasks:
		if len(parts) == 2 {
			return validAddressDate(parts[1])
		}
		if len(parts) == 3 {
			return validAddressDate(parts[1]) && pageSlugPattern.MatchString(parts[2])
		}
		return len(parts) == 4 && validAddressDate(parts[1]) &&
			pageSlugPattern.MatchString(parts[2]) && parts[3] == TaskTraceFile
	case LayerProjects, LayerWiki:
		if len(parts) != 2 {
			return false
		}
		slug, found := strings.CutSuffix(parts[1], MarkdownExtension)
		return found && pageSlugPattern.MatchString(slug)
	default:
		return false
	}
}

func validAddressDate(value string) bool {
	return value != "" && ValidateDate(value) == nil
}

func addressableSourceDocument(name string) bool {
	source, found := strings.CutSuffix(name, ".json")
	return found && sourceNamePattern.MatchString(source)
}

// Relative maps an absolute path inside the base back to its base-relative slash form. It is
// the inverse of Resolve and the only place a URI is minted from a filesystem walk.
func (s Store) Relative(absolute string) (string, error) {
	rel, err := filepath.Rel(s.root, absolute)
	if err != nil {
		return "", fmt.Errorf("%w: %s is outside %s", ErrPathEscapes, absolute, s.root)
	}
	slashed := filepath.ToSlash(rel)
	if slashed == ".." || strings.HasPrefix(slashed, "../") {
		return "", fmt.Errorf("%w: %s is outside %s", ErrPathEscapes, absolute, s.root)
	}
	return slashed, nil
}

// ConfigPath is the committed configuration file of this base.
func (s Store) ConfigPath() string { return filepath.Join(s.root, ConfigFileName) }

// LocalConfigPath is the gitignored machine-local overlay of this base.
func (s Store) LocalConfigPath() string { return filepath.Join(s.root, LocalConfigName) }

// configStore preserves the absolute root spelling the operator chose, including a root that
// is itself a symlink. That spelling is part of trust identity; only entries beneath it are
// confined and forbidden from being links.
func configStore(root string) (Store, error) {
	absolute, err := ResolveAbsolutePath(root)
	if err != nil {
		return Store{}, err
	}
	return Store{root: absolute}, nil
}

// readConfigLeaf is the one read boundary for both execution-defining YAML files. Existing
// leaves must be regular base-owned files: a symlink moves reviewed commands outside the base,
// while a FIFO or device can block or produce different bytes on each read. ValidateWithinRoot
// deliberately starts below store.root, so a symlink-spelled base root remains supported.
func readConfigLeaf(store Store, name string) ([]byte, bool, error) {
	if name != ConfigFileName && name != LocalConfigName {
		return nil, false, fmt.Errorf("%w: %s is not a configuration leaf", ErrUnsafePath, name)
	}
	path := filepath.Join(store.root, name)
	if err := ValidateWithinRoot(store.root, path); err != nil {
		return nil, false, err
	}
	info, err := os.Lstat(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil, false, nil
	}
	if err != nil {
		return nil, false, fmt.Errorf("%w: inspect %s: %w", ErrUnsafePath, path, err)
	}
	if info.Mode()&os.ModeSymlink != 0 || !info.Mode().IsRegular() {
		return nil, false, fmt.Errorf("%w: %s must be a regular non-symlink file", ErrUnsafePath, path)
	}
	data, err := ReadFileLimit(path, MaxConfigBytes)
	if err != nil {
		return nil, false, err
	}
	return data, true, nil
}

// BinDir is the base's own script directory, prepended to PATH for every declared command.
func (s Store) BinDir() string { return filepath.Join(s.root, BaseBinDir) }

// Versioned reports whether the base has recognizable, real git working-tree metadata.
// Detecting this beats declaring it: a configuration key could claim a base is versioned when
// it is not, and the permission contract would then be wrong in the one direction that matters.
func (s Store) Versioned() bool {
	marker := filepath.Join(s.root, ".git")
	info, err := os.Lstat(marker)
	if err != nil || info.Mode()&os.ModeSymlink != 0 {
		return false
	}
	if info.IsDir() {
		return hasGitHead(marker)
	}
	if !info.Mode().IsRegular() {
		return false
	}
	data, err := ReadFileLimit(marker, MaxControlFileBytes)
	if err != nil {
		return false
	}
	line := strings.TrimSpace(string(data))
	gitDir, found := strings.CutPrefix(line, "gitdir:")
	if !found || strings.ContainsAny(gitDir, "\r\n") {
		return false
	}
	gitDir = strings.TrimSpace(gitDir)
	if gitDir == "" {
		return false
	}
	if !filepath.IsAbs(gitDir) {
		gitDir = filepath.Join(s.root, gitDir)
	}
	return hasGitHead(filepath.Clean(gitDir))
}

func hasGitHead(gitDir string) bool {
	directory, err := os.Lstat(gitDir)
	if err != nil || !directory.IsDir() || directory.Mode()&os.ModeSymlink != 0 {
		return false
	}
	head, err := os.Lstat(filepath.Join(gitDir, "HEAD"))
	return err == nil && head.Mode().IsRegular() && head.Mode()&os.ModeSymlink == 0
}

// EnforcePermissions reports whether a permission audit may repair modes. A versioned base
// is only inspected.
func (s Store) EnforcePermissions() bool { return !s.Versioned() }

File: src/fkf/store.py
"""Published base paths, lexical addressing, and symlink confinement."""

from __future__ import annotations

import os
import posixpath
import re
import stat
from dataclasses import dataclass, field
from datetime import date
from enum import StrEnum
from pathlib import Path, PurePosixPath
from typing import Final

from fkf.errors import InvalidUsageError, OperationalError

BASE_DIR_MODE: Final = 0o700
BASE_FILE_MODE: Final = 0o600

GRAPH_FILE: Final = "graph.tsv"
GRAPH_DST_FILE: Final = "graph.dst.tsv"
GRAPH_OFFSETS_FILE: Final = "graph.offsets.tsv"
GRAPH_META_FILE: Final = "graph.meta.json"
GRAPH_GENERATION_FILE: Final = "graph.generation.json"
TASK_TRACE_FILE: Final = "TASKS.md"
BASE_AGENTS_FILE: Final = "AGENTS.md"
BASE_SKILLS_DIR: Final = ".agents/skills"
BASE_SOURCES_DIR: Final = "sources"
BASE_TESTS_DIR: Final = "tests"
BASE_CLIENTS_DIR: Final = "clients"
CONFIG_FILE_NAME: Final = "fkf.yaml"
LOCAL_CONFIG_NAME: Final = "fkf.local.yaml"
MARKDOWN_EXTENSION: Final = ".md"
BASE_ENV_VAR: Final = "FKF_BASE"
MAX_CONFIG_BYTES: Final = 1 << 20
MAX_CONTROL_FILE_BYTES: Final = 1 << 20
MAX_SOURCE_DOCUMENT_BYTES: Final = 64 << 20
MAX_LOCAL_INPUT_BYTES: Final = 64 << 20
MAX_NARRATIVE_BYTES: Final = 4 << 20

_PAGE_SLUG_PATTERN = re.compile(r"^[a-z0-9][a-z0-9._-]*$")
_SOURCE_NAME_PATTERN = re.compile(r"^[a-z0-9][a-z0-9-]*$")


class PathEscapesError(InvalidUsageError):
    """A purported base-relative path escapes its selected root."""


class UnsafePathError(OperationalError):
    """A path contains a symlink or unsafe filesystem entry."""


class NotAddressableError(InvalidUsageError):
    """A path is outside FKF's published resource grammar."""


class LayerDisabledError(InvalidUsageError):
    """A request addresses a layer disabled by this base."""

    def __init__(self, layer: Layer) -> None:
        self.layer = layer
        super().__init__(f"layer {layer} is disabled in {CONFIG_FILE_NAME}; set layers.{layer}: true to enable it")


class NoBaseError(InvalidUsageError):
    """No explicit, environment, or discovered base exists."""


class Layer(StrEnum):
    """One typed storage layer in a base."""

    EVENTS = "events"
    INDEX = "index"
    TASKS = "tasks"
    PROJECTS = "projects"
    WIKI = "wiki"


LAYERS: Final[tuple[Layer, ...]] = tuple(Layer)


def layer_names() -> str:
    """Render the canonical layer list."""
    return ", ".join(LAYERS)


def parse_layer(value: str) -> Layer:
    """Parse one case-insensitive layer name."""
    candidate = value.strip().lower()
    try:
        return Layer(candidate)
    except ValueError as error:
        raise ValueError(f"unknown layer {value!r}; valid layers: {layer_names()}") from error


def expand_home(value: str) -> str:
    """Expand exactly ``~`` and ``~/`` using this process's HOME."""
    cleaned = value.strip()
    if cleaned != "~" and not cleaned.startswith("~/"):
        return cleaned
    home = os.environ.get("HOME", "").strip()
    if not home:
        return cleaned
    return home if cleaned == "~" else str(Path(home) / cleaned[2:])


def resolve_absolute_path(value: str | os.PathLike[str]) -> Path:
    """Anchor a path without resolving its symlinks."""
    expanded = expand_home(os.fspath(value))
    if not expanded:
        raise ValueError("path is empty")
    # ``absolute`` preserves the chosen symlink spelling; trust identity depends on it.
    return Path(os.path.normpath(expanded)).absolute()


def resolve_physical_path(value: str | os.PathLike[str]) -> Path:
    """Resolve existing aliases while preserving a missing suffix."""
    current = resolve_absolute_path(value)
    missing: list[str] = []
    while True:
        try:
            resolved = current.resolve(strict=True)
        except FileNotFoundError:
            parent = current.parent
            if parent == current:
                raise OSError(f"resolve physical path {value}: no existing ancestor") from None
            missing.append(current.name)
            current = parent
            continue
        except OSError as error:
            raise OSError(f"resolve physical path {value}: {error}") from error
        for component in reversed(missing):
            resolved /= component
        return Path(os.path.normpath(resolved))


def clean_relative(relative: str) -> str:
    """Normalize a slash path and reject every spelling that can escape."""
    value = relative.strip()
    if not value:
        raise PathEscapesError("path escapes the base: path is empty")
    if "\0" in value or "\\" in value:
        raise PathEscapesError(f"path escapes the base: {value!r} contains an unsafe character")
    if value.startswith("/"):
        raise PathEscapesError(f"path escapes the base: {value!r} is absolute")
    if value.startswith("~"):
        raise PathEscapesError(f"path escapes the base: {value!r} is home-relative")
    trailing_slash = value.endswith("/")
    cleaned = posixpath.normpath(value)
    if cleaned == ".." or cleaned.startswith("../"):
        raise PathEscapesError(f"path escapes the base: {value!r}")
    # PurePosixPath also rejects nothing itself, so pin FKF's filesystem-neutral grammar.
    if any(part in {"", ".", ".."} for part in PurePosixPath(cleaned).parts):
        raise PathEscapesError(f"path escapes the base: {value!r} is not a valid path")
    return f"{cleaned}/" if trailing_slash and cleaned != "." else cleaned


def validate_date(value: str) -> None:
    """Require the exact YYYY-MM-DD shape, allowing an absent bound."""
    if not value.strip():
        return
    try:
        parsed = date.fromisoformat(value)
    except ValueError as error:
        raise ValueError(f"date must be YYYY-MM-DD: {error}") from error
    if parsed.isoformat() != value:
        raise ValueError("date must be YYYY-MM-DD")


def _root_alias_resolved(path: Path) -> Path:
    """Resolve only an OS-owned first component such as macOS ``/var``."""
    parts = path.parts
    if len(parts) < 2 or not path.is_absolute():
        return path
    root_entry = Path(parts[0], parts[1])
    try:
        entry = root_entry.lstat()
    except FileNotFoundError:
        return path
    except OSError as error:
        raise UnsafePathError(f"unsafe filesystem path: inspect {root_entry}: {error}") from error
    if not stat.S_ISLNK(entry.st_mode):
        return path
    try:
        resolved = root_entry.resolve(strict=True)
    except OSError as error:
        raise UnsafePathError(f"unsafe filesystem path: resolve system root alias {root_entry}: {error}") from error
    return resolved.joinpath(*parts[2:])


def _validate_path_confinement(path: str | os.PathLike[str], *, directory_leaf: bool) -> None:
    # Do not resolve symlinks before the component-by-component audit below.
    candidate = Path(os.path.normpath(expand_home(os.fspath(path)))).absolute()
    if os.fspath(path).strip() in {"", "."}:
        raise UnsafePathError("unsafe filesystem path: path is empty")
    candidate = _root_alias_resolved(candidate)
    current = Path(candidate.anchor)
    for component in candidate.parts[1:]:
        current /= component
        try:
            info = current.lstat()
        except FileNotFoundError:
            return
        except OSError as error:
            raise UnsafePathError(f"unsafe filesystem path: inspect {current}: {error}") from error
        if stat.S_ISLNK(info.st_mode):
            raise UnsafePathError(f"unsafe filesystem path: {current} is a symlink")
        if (current != candidate or directory_leaf) and not stat.S_ISDIR(info.st_mode):
            raise UnsafePathError(f"unsafe filesystem path: intermediate component {current} is not a directory")


def validate_path_confinement(path: str | os.PathLike[str]) -> None:
    """Reject existing symlinks and non-directory intermediate entries."""
    _validate_path_confinement(path, directory_leaf=False)


def validate_directory_confinement(path: str | os.PathLike[str]) -> None:
    """Also require an existing leaf to be a real directory."""
    _validate_path_confinement(path, directory_leaf=True)


def validate_within_root(root: str | os.PathLike[str], absolute: str | os.PathLike[str]) -> None:
    """Reject lexical escapes and every symlink strictly below a selected root."""
    base = Path(os.path.normpath(root))
    target = Path(os.path.normpath(absolute))
    try:
        relative = target.relative_to(base)
    except ValueError as error:
        raise PathEscapesError(f"path escapes the base: {target} is outside {base}") from error
    if relative == Path():
        return
    current = base
    for component in relative.parts:
        current /= component
        try:
            info = current.lstat()
        except FileNotFoundError:
            return
        except OSError as error:
            raise UnsafePathError(f"unsafe filesystem path: inspect {current}: {error}") from error
        if stat.S_ISLNK(info.st_mode):
            raise UnsafePathError(
                f"unsafe filesystem path: {current} is a symlink inside the base; a base addresses only its own files"
            )


def _valid_address_date(value: str) -> bool:
    try:
        validate_date(value)
    except ValueError:
        return False
    return bool(value)


def _addressable_source_document(name: str) -> bool:
    return name.endswith(".json") and bool(_SOURCE_NAME_PATTERN.fullmatch(name[:-5]))


def _addressable_layer_path(layer: Layer, relative: str) -> bool:
    parts = relative.split("/")
    if len(parts) == 1:
        return True
    if layer is Layer.EVENTS:
        if len(parts) == 2:
            return _valid_address_date(parts[1])
        return len(parts) == 3 and _valid_address_date(parts[1]) and _addressable_source_document(parts[2])
    if layer is Layer.INDEX:
        return len(parts) == 2 and _addressable_source_document(parts[1])
    if layer is Layer.TASKS:
        if len(parts) == 2:
            return _valid_address_date(parts[1])
        if len(parts) == 3:
            return _valid_address_date(parts[1]) and bool(_PAGE_SLUG_PATTERN.fullmatch(parts[2]))
        return (
            len(parts) == 4
            and _valid_address_date(parts[1])
            and bool(_PAGE_SLUG_PATTERN.fullmatch(parts[2]))
            and parts[3] == TASK_TRACE_FILE
        )
    if layer in {Layer.PROJECTS, Layer.WIKI}:
        return (
            len(parts) == 2
            and parts[1].endswith(MARKDOWN_EXTENSION)
            and bool(_PAGE_SLUG_PATTERN.fullmatch(parts[1][: -len(MARKDOWN_EXTENSION)]))
        )
    return False


def addressable_base_path(relative: str) -> bool:
    """Return whether a clean relative path belongs to the public grammar."""
    cleaned = relative.removesuffix("/")
    first = cleaned.partition("/")[0]
    try:
        layer = Layer(first)
    except ValueError:
        return cleaned in {
            GRAPH_FILE,
            GRAPH_DST_FILE,
            GRAPH_OFFSETS_FILE,
            GRAPH_META_FILE,
            GRAPH_GENERATION_FILE,
            CONFIG_FILE_NAME,
            BASE_AGENTS_FILE,
        }
    return _addressable_layer_path(layer, cleaned)


@dataclass(frozen=True, slots=True)
class Store:
    """Resolved immutable layout of one base."""

    root: Path
    _enabled: dict[Layer, bool] = field(repr=False)

    def __init__(self, root: str | os.PathLike[str], enabled: dict[Layer, bool] | None = None) -> None:
        object.__setattr__(self, "root", Path(os.path.normpath(expand_home(os.fspath(root)))))
        declared = enabled or {}
        object.__setattr__(self, "_enabled", {layer: bool(declared.get(layer, False)) for layer in LAYERS})

    def enabled(self, layer: Layer) -> bool:
        return self._enabled[layer]

    @property
    def enabled_layers(self) -> tuple[Layer, ...]:
        return tuple(layer for layer in LAYERS if self._enabled[layer])

    def directory(self, layer: Layer) -> Path:
        return self.resolve(layer)

    @staticmethod
    def layer_of(relative: str) -> Layer | None:
        first = relative.removeprefix("/").partition("/")[0]
        try:
            return Layer(first)
        except ValueError:
            return None

    def resolve(self, relative: str | Layer) -> Path:
        cleaned = clean_relative(str(relative)).removesuffix("/")
        if cleaned == ".":
            return self.root
        layer = self.layer_of(cleaned)
        if layer is not None and not self.enabled(layer):
            raise LayerDisabledError(layer)
        if not addressable_base_path(cleaned):
            raise NotAddressableError(
                f"path is not addressable in a base: {cleaned} "
                f"(a base addresses the published shapes under the {layer_names()} layers)"
            )
        absolute = self.root.joinpath(*cleaned.split("/"))
        validate_within_root(self.root, absolute)
        return absolute

    def relative(self, absolute: str | os.PathLike[str]) -> str:
        target = Path(os.path.normpath(absolute))
        try:
            return target.relative_to(self.root).as_posix()
        except ValueError as error:
            raise PathEscapesError(f"path escapes the base: {target} is outside {self.root}") from error

    @property
    def config_path(self) -> Path:
        return self.root / CONFIG_FILE_NAME

    @property
    def local_config_path(self) -> Path:
        return self.root / LOCAL_CONFIG_NAME

    @property
    def sources_dir(self) -> Path:
        return self.root / BASE_SOURCES_DIR

    @property
    def clients_dir(self) -> Path:
        return self.root / BASE_CLIENTS_DIR

    @property
    def tests_dir(self) -> Path:
        return self.root / BASE_TESTS_DIR

    @property
    def versioned(self) -> bool:
        marker = self.root / ".git"
        try:
            info = marker.lstat()
        except OSError:
            return False
        if stat.S_ISLNK(info.st_mode):
            return False
        if stat.S_ISDIR(info.st_mode):
            return _has_git_head(marker)
        if not stat.S_ISREG(info.st_mode):
            return False
        try:
            line = _read_small_regular(marker).decode().strip()
        except OSError, UnicodeError:
            return False
        if not line.startswith("gitdir:") or "\n" in line or "\r" in line:
            return False
        git_dir_value = line.removeprefix("gitdir:").strip()
        if not git_dir_value:
            return False
        git_dir = Path(git_dir_value)
        if not git_dir.is_absolute():
            git_dir = self.root / git_dir
        return _has_git_head(Path(os.path.normpath(git_dir)))

    @property
    def enforce_permissions(self) -> bool:
        return not self.versioned


def _read_small_regular(path: Path) -> bytes:
    before = path.lstat()
    if stat.S_ISLNK(before.st_mode) or not stat.S_ISREG(before.st_mode) or before.st_size > MAX_CONTROL_FILE_BYTES:
        raise UnsafePathError(f"unsafe filesystem path: {path} must be a bounded regular non-symlink file")
    with path.open("rb") as handle:
        after = os.fstat(handle.fileno())
        if not stat.S_ISREG(after.st_mode) or (before.st_dev, before.st_ino) != (after.st_dev, after.st_ino):
            raise UnsafePathError(f"unsafe filesystem path: {path} changed before it was opened")
        return handle.read(MAX_CONTROL_FILE_BYTES + 1)


def _has_git_head(git_dir: Path) -> bool:
    try:
        directory = git_dir.lstat()
        head = (git_dir / "HEAD").lstat()
    except OSError:
        return False
    return (
        stat.S_ISDIR(directory.st_mode)
        and not stat.S_ISLNK(directory.st_mode)
        and stat.S_ISREG(head.st_mode)
        and not stat.S_ISLNK(head.st_mode)
    )


def discover_base(explicit: str) -> tuple[Path, str]:
    """Select a base from flag, environment, then nearest ancestor."""
    if expanded := expand_home(explicit):
        try:
            return resolve_absolute_path(expanded), "flag"
        except ValueError as error:
            raise NoBaseError(f"no fkf base found: resolve --base {explicit!r}: {error}") from error
    environment = os.environ.get(BASE_ENV_VAR, "")
    if expanded := expand_home(environment):
        try:
            return resolve_absolute_path(expanded), "environment"
        except ValueError as error:
            raise NoBaseError(f"no fkf base found: resolve {BASE_ENV_VAR}={environment!r}: {error}") from error
    working = Path.cwd()
    for directory in (working, *working.parents):
        config = directory / CONFIG_FILE_NAME
        try:
            info = config.stat()
        except OSError:
            continue
        if stat.S_ISREG(info.st_mode):
            return directory, "discovery"
    raise NoBaseError(
        f"no fkf base found: pass --base <path>, export {BASE_ENV_VAR}, or run from inside a base "
        f"(a directory holding {CONFIG_FILE_NAME}, found by walking up from {working})"
    )


__all__ = [
    "BASE_AGENTS_FILE",
    "BASE_CLIENTS_DIR",
    "BASE_DIR_MODE",
    "BASE_ENV_VAR",
    "BASE_FILE_MODE",
    "BASE_SKILLS_DIR",
    "BASE_SOURCES_DIR",
    "BASE_TESTS_DIR",
    "CONFIG_FILE_NAME",
    "GRAPH_DST_FILE",
    "GRAPH_FILE",
    "GRAPH_GENERATION_FILE",
    "GRAPH_META_FILE",
    "GRAPH_OFFSETS_FILE",
    "LAYERS",
    "LOCAL_CONFIG_NAME",
    "Layer",
    "LayerDisabledError",
    "NoBaseError",
    "NotAddressableError",
    "PathEscapesError",
    "Store",
    "UnsafePathError",
    "addressable_base_path",
    "clean_relative",
    "discover_base",
    "expand_home",
    "layer_names",
    "parse_layer",
    "resolve_absolute_path",
    "resolve_physical_path",
    "validate_date",
    "validate_directory_confinement",
    "validate_path_confinement",
    "validate_within_root",
]
